server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -2065,6 +2067,7 @@ public boolean flush(boolean force, boolean waitIfOngoing) throws EngineExceptio
                     }
                     refreshLastCommittedSegmentInfos();
                     flushListener.afterFlush(lastCommittedSegmentInfos.getGeneration(), commitLocation);
+                    versionMap.afterFlush(lastCommittedSegmentInfos.getGeneration());
```

Comment:
I think this is too early, since the commit upload itself is done asynchronously. I wonder if we should move the responsibility of clearing the archived versions to the stateless index commit listener, which we can extend to notify upon upload completion?

Reply:
I suppose this *can* work out, if we then instead wait for the new commit to arrive in real-time GET. However, it would seem better to take that out of the equation and only clear this once the notification to search shards of the new commit completed.

Reply:
> which we can extend to notify upon upload completion?

Would we gain anything by pushing the `afterFlush` call to that moment? We'd have to still wait for the unpromotable shard refresh confirmation before clearing the archive map, I think. Once we have the generation of the commit, we could just send the unpromotable shard refresh actions and only if it returns successfully, clean up the archive map.

I believe we are doing also something similar to handle `WAIT_UNTIL` refresh policy requested by a bulk request, which is triggered by the line right above.

Reply:
If we fail to upload the commit, I think we do risk dirty-reads which we're fine with, it seems. But as far as pruning the archive map goes, either we wait until the node can upload the commit (since we wait for unpromotable refresh which waits for the commit notification) or the node fails eventually. i.e., confirmation from a search shard that it is on a generation (but from a different primary term) would prune the map, but at that point that indexing shard is not relevant anyway.

Reply:
> We'd have to still wait for the unpromotable shard refresh confirmation before clearing the archive map
Agreed, it is once the `NewCommitNotificationRequest` has succeeded.